src/token-builder/types.js

```diff
@@ -0,0 +1,192 @@
+const ALLOWED_ALGOS = new Set(['sha256', 'sha1', 'md5']);
+const ALLOWED_ENCODINGS = new Set(['hex', 'base64', 'base64url', 'base64percent']);
+
+export const REPLACE_CHAR_LIMIT = 100;
+
+export class TokenBase {
+  constructor({ name, cacheOverride = null, skipCache = false }) {
+    this.name = name;
+    this.type = 'base';
+    this.requiredProperties = ['name'];
+    this.errors = [];
+    this.cacheOverride = cacheOverride;
+    this.skipCache = skipCache;
+  }
+
+  validateOptions() {
+    const missingProps = [];
+    this.requiredProperties.forEach((prop) => {
+      if (!this[prop]) {
+        missingProps.push(prop);
+      }
+    });
+
+    if (missingProps.length) {
+      this.errors.push(`Missing properties for ${this.type} token: "${missingProps.join(', ')}"`);
+    }
+  }
+}
+
+export class ReplaceToken extends TokenBase {
+  constructor(params) {
+    super(params);
+    this.type = 'replace';
+    this.requiredProperties = [...this.requiredProperties, 'value'];
+    this.value = params.value;
+
+    this.validateOptions();
+  }
+
+  toJSON() {
```

Comment:
We used to have `toJSON` on `TokenBase` but removed it because a user wouldn't actually be instantiating a base token. Though should we still have that method on the base class?

Note: for the `toJSON` method on each token subclass, 
- `replace` and `replaceLarge` return  the same properties
- `hmac` and `sha1` return the same propertie
- `secret` token is unique